### ISSUE TITLE
[labs/ssr] Can't render light DOM custom elements server side

### ISSUE THREAD

**Issue by eddyloewen** [lit/lit#1994]
# Description

I'm trying to SSR custom elements that render to light DOM. Everything seems to work correctly when rendering to shadow DOM.

# Steps to Reproduce

1. I'm not using LitElement as my base class. I use https://github.com/webtides/element-js/. I only use lit-html for rendering. And I do render to light DOM by default.

```ts
import {render} from '@lit-labs/ssr/lib/render-with-global-dom-shim.js';
import {TemplateElement} from '@webtides/element-js';
import {html} from "lit-html";

class MyRenderer extends ElementRenderer {/*...*/}

class LightElement extends TemplateElement {
    template() {
        return html`
            <div>Parent light DOM element</div>
            <nested-light-element></nested-light-element>
        `;
    }
}
customElements.define('light-element', LightElement);

class NestedLightElement extends TemplateElement {
    template() {
        return html`
            <div>Nested light DOM element</div>
        `;
    }
}
customElements.define('nested-light-element', NestedLightElement);

const template = html`
   <light-element></light-element>
`;

render(template, {
    elementRenderers: [MyRenderer],
    customElementInstanceStack: [],
    customElementHostStack: [],
})
```

2. Everything seems to work fine at first. Both elements (light-element and nested-light-element) are rendered on the server. Both elements get hydraded correctly. I can update the nested element (nested-light-element) by changing properties or attributes and it will re-render. But when I update the parent element (light-element) by changing properties or attributes it will destroy the nested element's template and it won't recover from that. The parent element will re-render just fine but the nested element can't ever render again.
3. When I opt into rendering to shadow DOM instead, everything works as expected.
4. From what I can tell by looking at the source (@lit-labs/ssr/lib/render-lit-html.js:300), it will always assume shadow DOM. There is no TemplateOpcode for light DOM.
5. Looking further in the source (@lit-labs/ssr/lib/render-lit-html.js:498), it will always wrap the template with a declarative shadow root template element. There seems to be no case for rendering light DOM server side here.
6. Also the `renderLight` method in my `ElementRenderer` never gets called. Wouldn't it make sense to let the ElementRenderer decide whether to render to light DOM or to shadow DOM?

## Expected Results

I can SSR nested custom elements that render to light DOM.

## Actual Results

Nested custom element's templates break when rendered to light DOM

**Comment by kevinpschaaf** [comment]
You're right, currently the LitElement ElementRenderer only supports rendering to shadow root; we haven't made a pattern for `render-lit-html` render `LitElements` that render directly to their light DOM.

While we see users find utility in rendering an element's template to its light DOM (i.e. in LitElement, by overriding `createRenderRoot`), this is not a generally interoperable pattern because it leaves undefined what happens with children that are rendered by the element's owner to its light DOM (e.g. do they stay? are they overwritten? what order do they go in w.r.t. the DOM the CE renders into itself? what happens when additional nodes are added to the light DOM after the initial render?).

```html
<!-- This node belongs to the outer scope -->
<some-host-element>
  <!-- This node _also_ belongs to the outer scope -->
  <div>outer scope</div>
  <!-- These nodes belong to some-host-element -->
  <div>rendered by some-host-element</div>
  <!-- Is this the right order? What if we add more? Lots of questions... -->
</some-host-element>
```

The design of web components (the combination of both custom elements _and_ shadow DOM) make the scope of node ownership explicit and interoperable:

```html
<!-- This node belongs to the outer scope -->
<some-host-element>
  <!-- Everything in the light DOM also belongs to the outer scope -->
  <div>outer scope</div>
  #shadow-root
    <!-- These nodes belong to some-host-element -->
    <div>rendered by some-host-element</div>
    <slot><!--controls where light DOM goes--></slot>
  #/shadow-root
</some-host-element>
```

We've speculatively added `renderLight` to the `ElementRenderer` interface to explore a better pattern for "light DOM rendering" than having elements render directly to their light DOM. For Lit, it goes along with the `renderLight` directive found in `@lit-labs/ssr-client`.  We need to add better examples and documentation to help explain how we see it used and to get feedback on it.

The design of `renderLight` helps to avoid ambiguity around the ownership of elements that comes along with light-DOM rendering. The `some-host-element` no longer autonomously renders nodes to the outer scope which is technically not in its scope of ownership; its `renderLight` template is only rendered when the owner of some-host-element uses the `renderLight` directive to tell Lit where it wants the light DOM children rendered:

```js
html`
  <!-- This node belongs to the outer scope -->
  <some-host-element>
    <!-- This node _also_ belongs to the outer scope -->
    <div>outer scope</div>
    <!-- Causes some-host-element's renderLight template to be rendered to its light DOM in this spot -->
    ${renderLight()}
</some-host-element>
```
In this way, the ownership is clear, in that the outer scope always controls what gets rendered to its scope; `renderLight` becomes an explicit delegation mechanism from the outer scope to the inner scope, which removes the ambiguity. Effectively, the outer scope "asks" the custom element to render some child nodes to a spot of the outer scope's choosing. While this affects how users _use_ the custom element (it's no longer fully autonomous), we think this is a better architectural approach that does not violate principles of ownership and source-of-truth in the DOM tree.

All that said, while we'd much rather fully embrace the interoperability provided by Shadow DOM and design light DOM rendering patterns on top of that, we could revisit whether there could be an alternate signal to automatically call `renderLight` on the `ElementRenderer` rather than require the outer scope to always use a signal like the `renderLight` directive.
